nvflare/fuel/utils/config.py

```diff
@@ -0,0 +1,192 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Dict, List, Optional
+
+
+class ConfigFormat(Enum):
+    JSON = "JSON"
+    PYHOCON = "PYHOCON"
+    OMEGACONF = "OMEGACONF"
+
+    @classmethod
+    def _config_format_extensions(cls):
+        return {
+            "JSON": [".json", ".json.default"],
+            "PYHOCON": [".conf", ".conf.default"],
+            "OMEGACONF": [".yml", ".yml.default"],
+        }
+
+    @classmethod
+    def search_order(cls):
+        return [cls.JSON, cls.PYHOCON, cls.OMEGACONF]
+
+    @classmethod
+    def get_extensions(cls, fmt: str) -> Optional[List[str]]:
+        return cls._config_format_extensions().get(fmt, None)
+
+    @classmethod
+    def config_exts(cls) -> str:
+        search_sequence: List[ConfigFormat] = cls.search_order()
+        extensions = []
+        for fmt in search_sequence:
+            exts = cls.get_extensions(fmt.name)
+            if exts:
+                extensions.extend(exts)
+
+        return "|".join(iter(extensions))
+
+
+class Config(ABC):
+    @abstractmethod
+    def get_native_conf(self):
+        """
+        Return the original underline config object representation if you prefer to use it directly
+        Pyhocon → ConfigTree
+        JSON → Dict
+        OMEGACONF → ConfigDict
+
+        Returns: Any, native config objects
+        """
+
+        pass
+
+    @abstractmethod
+    def get_format(self):
+        """
+        Returns: ConfigFormat, returns the current config objects ConfigFormat
+        """
+
+        pass
+
+    @abstractmethod
+    def get_location(self) -> Optional[str]:
+        """
+        Returns: None if the config is not from file else return the file path where this configuration is loaded from
+
+        """
+        pass
+
+    @abstractmethod
+    def to_dict(self) -> Dict:
+        """
+        convert underline config object to dictionary
+        Returns:Dict
+
+        """
+        pass
+
+    @abstractmethod
+    def to_conf(self, element: Dict) -> str:
+        """
+            convert dict element to the str representation of the underline configuration.
+            For example, for JsonFormat, the method return json string
+            for PyhoconFormat, the method return pyhocon string
+            for OmegaconfFormat, the method returns YAML string representation
+        Args:
+            element: dict
+
+        Returns: str string representation of the configuration in given format
+
+        """
+
+        pass
+
+    @abstractmethod
+    def get_int(self, key: str, default=None) -> Optional[int]:
```

Comment:
Why do we need these get_xxx() methods?

I thought the conf loader will only return a dict. It will be up to another layer (i.e. ConfigService) to do all these get_xxx() stuff.

Reply:
^ "the get_xxx just enforce the types or interpret the types."

So you mean for the same key say "num_of_rounds", we could use config.get_int("num_of_rounds") to get an integer representation and we could also use config.get_float("num_of_rounds") to get as float?

Then the ConfigService's get_xxx will try to 
1. get this key from environment variable
2. get this key from config_fed_server.json using "config.get_xxx"
3. get this key from resources.json using "config.get_xxx"
4. get this key from command line arguments

^ the order can be what we decide.
The first value return will be the value that is used.

Is this what we are thinking?

Reply:
Yes and no. Even though each config could provide its own get_xxx(), it's unnecessary since everything is loaded into dict already, and all vars could be retrieved from the dict, regardless where the dict is loaded from.

Secondly, as to how the var's value is determined, it could vary, depending on the nature of the var. What you laid out here is just one possible path. There are other cases.

Bottom line is that the var value determination logic is beyond a specific config loader.